src/main/java/uk/ac/sanger/sccp/stan/service/VisiumPermDataService.java

```diff
@@ -38,40 +45,87 @@ public VisiumPermDataService(LabwareRepo lwRepo, MeasurementRepo measurementRepo
+                .map(ss -> ss.getSlot().getId())
+                .collect(toSet());
         List<Measurement> measurements = measurementRepo.findAllBySlotIdIn(slotIds);
-        List<AddressPermData> pds = compilePermData(lw, measurements);
+        List<AddressPermData> pds = compilePermData(measurements, ssToAddress);
         return new VisiumPermData(lw, pds);
     }
 
+    /**
+     * Makes a map of ancestral slot and sample id to ultimate address in the given labware
+     * @param lw the labware
+     * @param ancestry the ancestry of this labware
+     * @return a map of ancestral slot and sample id to address in the given labware
+     */
+    public Map<SlotIdSampleId, Set<Address>> makeSlotSampleIdAddressMap(Labware lw, Ancestry ancestry) {
+        final Map<SlotIdSampleId, Set<Address>> ssToAddress = new HashMap<>();
+        final TriConsumer<Slot, Sample, Address> addToMap = (slot, sample, address) -> {
+            ssToAddress.computeIfAbsent(new SlotIdSampleId(slot, sample), k -> new HashSet<>()).add(address);
+        };
+
+        for (Slot slot : lw.getSlots()) {
+            for (Sample sam : slot.getSamples()) {
+                addToMap.accept(slot, sam, slot.getAddress());
+            }
+        }
+
+        Ancestoriser.SlotSample.stream(lw).forEach(targetSs -> {
+            Address targetAddress = targetSs.getSlot().getAddress();
+            for (Ancestoriser.SlotSample ancesterSs : ancestry.ancestors(targetSs)) {
+                addToMap.accept(ancesterSs.getSlot(), ancesterSs.getSample(), targetAddress);
+            }
+        });
+        return ssToAddress;
+    }
+
     /**
      * Converts measurements to perm data
-     * @param lw the labware the measurements are recorded on
      * @param measurements the measurements to convert
+     * @param ssToAddress a map of slot and sample ids to address to indicate in the data
      * @return perm data of the given measurements
      */
-    public List<AddressPermData> compilePermData(Labware lw, Collection<Measurement> measurements) {
-        Map<Integer, Address> slotIdToAddress = lw.getSlots().stream()
-                .collect(toMap(Slot::getId, Slot::getAddress));
-        Measurement selected = null;
+    public List<AddressPermData> compilePermData(Collection<Measurement> measurements,
+                                                 Map<SlotIdSampleId, Set<Address>> ssToAddress) {
         final LinkedHashSet<AddressPermData> pds = new LinkedHashSet<>(measurements.size());
-        Iterable<Measurement> sortedMeas = () -> measurements.stream()
+        List<Measurement> sortedMeasurements = measurements.stream()
                 .sorted(Comparator.comparing(Measurement::getId))
-                .iterator();
-        for (Measurement meas : sortedMeas) {
+                .collect(toList());
+
+        Measurement selectedMeasurement = null;
+        for (Measurement meas : reverseIter(sortedMeasurements)) {
```

Comment:
Yes. If there happen to be multiple _selected time_ measurements, we want to assume the latest one is the relevant one. Since the measurements are sorted lowest to highest, we can find the latest one by starting at the end of the measurements list and stopping at the first one we find.

Reply:
Cool, Thank you